src/bootstrap/download.rs

```diff
@@ -367,26 +367,69 @@ impl Config {
 
     pub(crate) fn download_ci_rustc(&self, commit: &str) {
         self.verbose(&format!("using downloaded stage2 artifacts from CI (commit {commit})"));
+
         let version = self.artifact_version_part(commit);
+        // download-rustc doesn't need its own cargo, it can just use beta's. But it does need the
+        // `rustc_private` crates for tools.
+        let extra_components = ["rustc-dev"];
+
+        self.download_toolchain(
+            &version,
+            "ci-rustc",
+            commit,
+            &extra_components,
+            Self::download_ci_component,
+        );
+    }
+
+    pub(crate) fn download_beta_toolchain(&self) {
+        self.verbose(&format!("downloading stage0 beta artifacts"));
+
+        let date = &self.stage0_metadata.compiler.date;
+        let version = &self.stage0_metadata.compiler.version;
+        let extra_components = ["cargo"];
+
+        let download_beta_component = |config: &Config, filename, prefix: &_, date: &_| {
+            config.download_component(DownloadSource::Dist, filename, prefix, date, "stage0")
+        };
+
+        self.download_toolchain(
+            version,
+            "stage0",
+            date,
+            &extra_components,
+            download_beta_component,
+        );
+    }
+
+    fn download_toolchain(
+        &self,
+        channel: &str,
```

Comment:
On stable, this channel is going to look like "1.67.1" or similar. So it's closer to *version*, not channel.

Is there a reason not to pass the config::CompilerMetadata struct into download_toolchain?

Reply:
> On stable, this channel is going to look like "1.67.1" or similar. So it's closer to version, not channel.

Yes, I see it's already passed as variable called `version` from other functions.

> Is there a reason not to pass the config::CompilerMetadata struct into download_toolchain?

`config::CompilerMetadata` isn't always used for passing `version`. In `download_ci_rustc`, `artifact_version_part` is used for calculating `version` which gets passed into `download_toolchain`.

Reply:
I think it would be nice to refactor to always pass in CompilerMetadata, since it seems like we pass both of those fields here. That would also avoid having three &str arguments side-by-side which can be easy to flip accidentally (and I'm not sure we have good test coverage, so we might not catch that flip).

Reply:
I fixed the previous review notes because they were pretty simple. Do you want me to add `FIXME` on top of the `version` arg, or do it in this PR? It requires some effort and I will not have enough space on this task for 3-4 days.

Reply:
I would prefer to do it in this PR, but no strong feelings. r=me if you want to move ahead with a FIXME.

Reply:
Let's not block this PR for that reactoring.

Just left a `FIXME(ozkanonur)` for it.